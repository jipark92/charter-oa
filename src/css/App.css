div {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

table {
  margin: 1rem;
}