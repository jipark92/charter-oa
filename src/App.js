import './css/App.css';
import MayMonthTable from './components/MayMonthTable';
import JuneMonthTable from './components/JuneMonthTable';
import JulyMonthTable from './components/JulyMonthTable';
import { useIsLoadingAndFetchData } from './hooks/useIsLoadingAndFetchData'

function App() {

  //custom hook loading and fetch 
  const { isLoading, transactionMockData } = useIsLoadingAndFetchData()

  //categorized table array
  const mayTable = []
  const juneTable = []
  const julyTable = []

  //find may data
  const findMayDateAndCustomerID = (() => {
    transactionMockData.map((data) => {
      if (data.customerID && new Date(data.purchasedDate).getMonth() + 1 === 5) {
        mayTable.push(data)
      }
    })
  })()
  const sortedMayTable = () => mayTable.sort((a, b) => new Date(a.purchasedDate) - new Date(b.purchasedDate))

  //find june data
  const findJuneDateAndCustomerID = (() => {
    transactionMockData.map((data) => {
      if (data.customerID && new Date(data.purchasedDate).getMonth() + 1 === 6) {
        juneTable.push(data)
      }
    })
  })()
  const sortedJuneTable = () => juneTable.sort((a, b) => new Date(a.purchasedDate) - new Date(b.purchasedDate))

  //find july data
  const findJulyDateAndCustomerID = (() => {
    transactionMockData.map((data) => {
      if (data.customerID && new Date(data.purchasedDate).getMonth() + 1 === 7) {
        julyTable.push(data)
      }
    })
  })()
  const sortedJulyTable = () => julyTable.sort((a, b) => new Date(a.purchasedDate) - new Date(b.purchasedDate))

  //calcualte may total reward points
  const calculateMayTotalRewards = (customer) => {
    const mayRewardPointsArr = []
    mayTable.map((data) => {
      if (data.customerID === customer) {
        mayRewardPointsArr.push(data.rewardPoints)
      }
    })
    return mayRewardPointsArr.reduce((total, pointsArr) => {
      return total + pointsArr
    }, 0)
  }
  //calcualte june total reward points
  const calculateJuneTotalRewards = (customer) => {
    const juneRewardPointsArr = []
    juneTable.map((data) => {
      if (data.customerID === customer) {
        juneRewardPointsArr.push(data.rewardPoints)
      }
    })
    return juneRewardPointsArr.reduce((total, pointsArr) => {
      return total + pointsArr
    }, 0)
  }
  //calcualte july total reward points
  const calculateJulyTotalRewards = (customer) => {
    const julyRewardPointsArr = []
    julyTable.map((data) => {
      if (data.customerID === customer) {
        julyRewardPointsArr.push(data.rewardPoints)
      }
    })
    return julyRewardPointsArr.reduce((total, pointsArr) => {
      return total + pointsArr
    }, 0)
  }

  return (
    <div className="App">
      {!isLoading ?
        <section>
          <MayMonthTable
            sortedMayTable={sortedMayTable}
            calculateMayTotalRewards={calculateMayTotalRewards}
          />
          <JuneMonthTable
            sortedJuneTable={sortedJuneTable}
            calculateJuneTotalRewards={calculateJuneTotalRewards}
          />
          <JulyMonthTable
            sortedJulyTable={sortedJulyTable}
            calculateJulyTotalRewards={calculateJulyTotalRewards}
          />
        </section> : <p>LOADING DATA......</p>}
    </div>
  );
}

export default App;
