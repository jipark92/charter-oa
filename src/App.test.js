import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';
import App from './App';

describe('App', () => {

    it('should match snapshot', () => {
        const { asFragment } = render(<App />)
        expect(asFragment()).toMatchSnapshot()
    })

    it("renders loading data text", () => {
        render(<App />)
        const loadingData = screen.getByTestId('loading-data').textContent
        expect(loadingData).toBe('LOADING DATA......')
    })

    it("should render components may,june,july table after 1 second", async () => {
        render(<App />)
        await waitFor(() => {
            let appComponents = screen.getByTestId('app-components')
            expect(appComponents).toBeInTheDocument('<MayMonthTable />')
        })
        await waitFor(() => {
            let appComponents = screen.getByTestId('app-components')
            expect(appComponents).toBeInTheDocument('<JuneMonthTable />')
        })
        await waitFor(() => {
            let appComponents = screen.getByTestId('app-components')
            expect(appComponents).toBeInTheDocument('<JulyMonthTable />')
        })
    })

    it('loading should change from true to false', async () => {
        const { isLoading } = render(<App />)
        expect(isLoading).not.toBe(true)
    })
})