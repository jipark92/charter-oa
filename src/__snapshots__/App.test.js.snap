// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <p
      data-testid="loading-data"
    >
      LOADING DATA......
    </p>
  </div>
</DocumentFragment>
`;
